lib/pbench/server/database/models/users.py

```diff
@@ -15,20 +15,28 @@ class Roles(enum.Enum):
     ADMIN = 1
 
 
+class UserProfileBadFormat(Exception):
+    """A base class for errors reported by the User class."""
+
+    def __init__(self, key: str):
+        self.key = key
+
+    def __str__(self) -> str:
+        return f"Generic error on {self.dataset} key {self.key}"
+
+
 class User(Database.Base):
     """User Model for storing user related details."""
 
     __tablename__ = "users"
 
     id = Column(Integer, primary_key=True, autoincrement=True)
+    oidc_id = Column(String(255), unique=True, nullable=False)
     username = Column(String(255), unique=True, nullable=False)
-    first_name = Column(String(255), unique=False, nullable=False)
-    last_name = Column(String(255), unique=False, nullable=False)
-    password = Column(LargeBinary(128), nullable=False)
     registered_on = Column(DateTime, nullable=False, default=datetime.datetime.now())
     email = Column(String(255), unique=True, nullable=False)
     role = Column(Enum(Roles), unique=False, nullable=True)
-    auth_tokens = relationship("ActiveTokens", backref="users")
+    profile = Column(JSON, unique=False, nullable=True)
```

Comment:
Note that `unique=False, nullable=True` are the default values, so we really don't need them here. (Or in `role`, if we really want to retain *that* column; given our discussions around the "RDU2 shared realm" I think we will want to rely on Keycloak roles defined lab-wide across multiple services rather than managing our own.)

How do we expect that `email` is populated? Is that in the OIDC decrypted payload, or are you just suggesting making that part of an "extended Pbench profile"? (And, if so, do we really want this as a separate column or simply a convention in your `profile` JSON?)

And `registered_on` is now, at best, the first time a Keycloak identity was used to authenticate to the Pbench server. A "first encounter", which I suppose could be used to managed some local cache purge if we want... but maybe the name promises too much?

Also, as with the API key table, I'm not sure that the autoincrement `id` is really doing us any favors here now that we're dropping the whole foreign key and relationship stuff. It's not anything special to SQL.

Reply:
> in `role`, if we really want to retain _that_ column; given our discussions around the "RDU2 shared realm" I think we will want to rely on Keycloak roles defined lab-wide across multiple services rather than managing our own.)

Yes but we also talked about having `put` specific tokens for example and that is something specific to the Pbench. User use the OIDC access token to generate the new API key where we might include this role in the key that later tells us that the key can only be used for `put`. Or we could have a generic role where if it is assigned to an API key it can be used to access other resources on Pbench not just put, I don't know. So removing the `role` column at this point seemed a little premature so I kept it. Maybe we can have a role as a part of the `profile` but I didn't want to spend too much time thinking about how its gonna look eventually.  

> How do we expect that `email` is populated? Is that in the OIDC decrypted payload

Yes, we get the email string from the decoded tokens.

> And `registered_on` is now, at best, the first time a Keycloak identity was used to authenticate to the Pbench server. A "first encounter", which I suppose could be used to managed some local cache purge if we want... but maybe the name promises too much?

Yeah, but I didn't think it promises too much, I mean this is an internal user table and `registered_on` in our table should mean when we encounter the user for the first time, isn't it? It has nothing to do with OIDC.


> Also, as with the API key table, I'm not sure that the autoincrement `id` is really doing us any favors here now that we're dropping the whole foreign key and relationship stuff. It's not anything special to SQL.

Make sense.

Reply:
How do we reconcile Keycloak and local private roles? Do we really want to build our own API and DB to support private roles? And even if we do, there's a lot more infrastructure to make that happen than just the column here, so I'm not convinced we ought to retain it. *IF* we add private roles, we'll add something like it then.

As for email... yeah, OK, but how much context do we want to capture locally beyond the username <-> ID we actually need to work with Keycloak? I guess we either capture everything that's in the OIDC token, or we capture only what we immediately need and think about adding more later.

Reply:
I have addressed it in the latest commit. I am only keeping internal `id`, `oidc_id`, and `username` and everything else that we extract from the token will go to the `profile` key.

Reply:
> I really cannot see why the Pbench Server ought to bother retaining _my_ name and email when it only lets _me_ see it. I kinda know that stuff already ...

Users may be able to update them on the Pbench server if they wish to have a different email for example on the Pbench server. The use case I can think about retaining these fields is from the dashboard perspective, when the user repeatedly logs in to the dashboard, instead of relying on `userinfo` endpoint of the OIDC server, the dashboard can rely on the pbench-server to get this information. From the server point of view, it doesn't add much value I agree.

> I don't see a use-case for "user private" profile information; even aside from the fact that unless we make the user profile "public", _everything_ we put there is private anyway. There is a use case for user profile attributes like "default retention period", which could be either columns or JSON.

My initial thoughts behind the `profile` column were that we can keep the `roles` information stored in the column (as a JSON format) which the user can access but can not change. So the profile column as it currently stands will have the `user` namespace to store anything that we could use to help display that information on the dashboard and its updatable whereas `server` namespace holds information like registered_on/roles which are not updatable. Both fields are private from the authorization point of view.

Reply:
If you want a "get user info" API on the server for PII that's private to the authenticated user, that's easy: you pass your API token into an API that simply decrypts that token and returns the information directly. We don't have to maintain a local cache and keep it up to date, and we want the API to be authenticated anyway, so we need to decrypt the token to check that they can get it ... and at that point it's right there in front of us without a DB query...

Reply:
Okay, that does make sense, we can decrypt the tokens and get the user information without making any additional db query. Alright, I going to revert these changes back and keep it very simple in the User table. I'll probably drop the profile column we don't need to store any user-related information at this point and we don't have a proper use case to do so. I am thinking to keep the `email` as a DB column since I can see the use of the email field for many reasons. `registered_on`/`created` might be good for auditing and I will keep the `role` the way it was before because removing that will break the `is_admin` functionality and I should raise a separate PR to fix that.

Reply:
Changing it back to draft until I make these changes.

Reply:
> I am thinking to keep the `email` as a DB column since I can see the use of the email field for many reasons.

You can "see" them...but they are still far away...I would drop it now.  It's not hard to add it later when we actually have a use for it.  (I'll repeat my earlier questions:  if we `had` an email address for a user, what would we use it for?  Who would send the email and why?)

> `registered_on`/`created` might be good for auditing

For auditing, we'll have the audit log where we can look it up if we really care.  (I can't think of any date-based operations that we would want to perform.  E.g., I think we're going to want to retain the UUID-to-username translation for at least as long as we have datasets owned by the UUID...and the registration date has nothing to do with that.)

> I will keep the `role` the way it was before because removing it will break the `is_admin` functionality and I should raise a separate PR to fix that.

Sounds OK to me.  (I still haven't grokked how we're going to implement `is_admin`....)

Reply:
> You can "see" them...but they are still far away...I would drop it now. It's not hard to add it later when we actually have a use for it. (I'll repeat my earlier questions: if we `had` an email address for a user, what would we use it for? Who would send the email and why?)
> 
I was thinking that if someone finds a public dataset of a particular interest and wants to contact the owner of the dataset, and if the owner of the dataset has a public email address attached to it, they would be able to communicate with each other.



> For auditing, we'll have the audit log where we can look it up if we really care. (I can't think of any date-based operations that we would want to perform. E.g., I think we're going to want to retain the UUID-to-username translation for at least as long as we have datasets owned by the UUID...and the registration date has nothing to do with that.)

When I said auditing I meant with the `registered_on` field we can see how many new users started using the Pbench server and what timeframe we get that new traffic, without this field we wouldn't be able to search the User table to determine whether there is a particular month that we see increased traffic of the new users. We can understand how many new users are using the Pbench each month or year and so.

Reply:
> I was thinking that if someone finds a public dataset of a particular interest and wants to contact the owner of the dataset, and if the owner of the dataset has a public email address attached to it, they would be able to communicate with each other.

I agree that that would be a great feature...but I don't think we should try to offer it between now and mid-March.  (And, I think that making it work will require lots of niggly bits -- the owner of the dataset has to be able to provide his/her email and know that it is publicly available, the Server has to store it, the reader of the dataset has to be able to fetch it, we need API support and Dashboard support, and so forth...none of which gets us a deployed Server faster.)

> with the registered_on field we can see how many new users started using the Pbench server and what timeframe we get that new traffic

True; but (hopefully) we should be able to glean that information from the audit log, as well.  This information will be write-once/read-maybe data which the Server itself is unlikely to ever use, so there is little point to maintaining it in the Server's database.  Having it in the `User` database table might make it somewhat easier to access, but depending on how the audit log is represented (e.g., in Elasticsearch or some other database), it might be just as easy to access.  And, even if it's just in a flat file, for the number of times that we need to generate this report, we can afford the added expense (if it's even noticeable).

I think it would be best to hold the `Users` table to the minimum, at least through Server 1.0....